### DISCUSSION TITLE
Is it possible to prevent file creation on specific paths by suffix?

### DISCUSSION THREAD

**Discussion by osevan** [netblue30/firejail#5151] [Q&A]
I have additonal hardening enhancement for firejailed apps like readonly mounting but more fine granular.

Imagine you want read only mount on some paths,  but prevents writing only by suffix to write  in paths.

Example :

/html directory needed to be readonly,  BUT only for php suffix like *.php, and everything else are allowed to write, only prevents .php .exe  .sh files for write in this path.

At the moment NO application can do that.

Maybe firejail can protect paths like above. ..

**Comment by rusty-snake** [comment]
> /html directory needed to be readonly, BUT only for php suffix like *.php, and everything else are allowed to write, only prevents .php .exe .sh files for write in this path.

You mean something like

```
read-only /html/*.php
read-only /html/*.exe
read-only /html/*.sh
```
?

**Reply by osevan** [reply]
Jup exactly !!!!

And prevent renaming in these suffix php sh exe.

Cause apps are allowed to create html files but renaming should not be allowed in php sh and exe file suffix

**Reply by rusty-snake** [reply]
> And prevent renaming in these suffix php sh exe.

This is not possible with mount-namespaces (nor with landlock).

**Reply by osevan** [reply]
> This is not possible with mount-namespaces (nor with landlock).

I know,  but you can work around in firejail like a function directoryprotect() and try to avoid file suffix writing.

**Reply by osevan** [reply]
Because this is big hole in Linux,  no one already wrote application for this type of service

**Reply by osevan** [reply]
Here is a example of monitoring filesystem events:https://man7.org/linux/man-pages/man7/inotify.7.html
Only a idea how to figure out if some changes to path happen, and if yes deleting files as fast as possible in firejail function.

**Comment by osevan** [comment]
Here bash solution for PDF files with inotifywait tool.

https://stackoverflow.com/questions/46542819/bash-script-inotifywait-wait-for-file-completely-written-before-proceding

**Reply by rusty-snake** [reply]
This would be racy, i.e. a false sense of security.

**Reply by osevan** [reply]
But possible for php files

**Reply by rusty-snake** [reply]
Is it racy? Then it is a false sense of security.

What do you mean with "possible for php files"?

**Reply by osevan** [reply]
> 

Its safe for usecases like preventing php files for executing or sh files. ... because firejail inotify monitoring level are faster in deleting as a human execute over webserver

**Comment by rusty-snake** [comment]
BTW, you know that file-extensions do not really matter in Linux.

```console
$ echo '#!/bin/sh\necho "Hello World"' > hello_world.py
$ file hello_world.py 
hello_world.py: POSIX shell script, ASCII text executable
```

**Reply by osevan** [reply]
I know and still big security hole

**Comment by osevan** [comment]
It is safe to monitor every 500ms if php exe or sh files created with Firejail function inside Sandbox and delete file.

**Reply by rusty-snake** [reply]
It is not. 500ms are years if you talk about modern cpus.

**Reply by osevan** [reply]
Try to write own thread for monitoring to avoid blocks /locking

**Comment by topimiettinen** [comment]
FUSE could be used to serve the files very selectively and there wouldn't be race issues. I proposed something similar for [systemd](https://lists.freedesktop.org/archives/systemd-devel/2015-March/028920.html). It would have allowed even disk block size granularity for access control.

**Reply by rusty-snake** [reply]
Yes, FUSE or seccomp-unotify can be used for that but you need to mirror hunders LoC from the kernel. IMHO absolute overkill.

In more recent kernel versions the fanotify API is maybe capable for this kind for filesystem event interception (I didn't look closer if it really can interception this case).
